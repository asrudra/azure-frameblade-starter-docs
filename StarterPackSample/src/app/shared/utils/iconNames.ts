// The full list of Fabric icons can be found here: https://developer.microsoft.com/en-us/fabric#/styles/icons
export enum IconNames {
    Allow = 'SkypeCircleCheck',
    Block = 'BlockedSolid',
    Delete = 'Delete',
    Save = 'Save',
    TestCase = 'TestCase',
    Undo = 'Undo'
}
