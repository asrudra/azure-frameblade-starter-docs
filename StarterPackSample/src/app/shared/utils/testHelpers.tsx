import * as React from 'react';
import { shallow } from 'enzyme';

export const testWithLocalizationContext = (Target: JSX.Element, enzymeWrapper: any = shallow) => { // tslint:disable-line:no-any
    const outerWrapper = shallow(Target); // tslint:disable-line:no-any
    const Children = outerWrapper.props().children;
    return enzymeWrapper(<Children t={jest.fn()}/>);
};
