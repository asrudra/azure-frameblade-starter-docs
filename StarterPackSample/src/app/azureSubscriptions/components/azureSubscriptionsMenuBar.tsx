import * as React from 'react';
import { CommandBar, ICommandBarItemProps } from 'office-ui-fabric-react/lib/CommandBar';
import { TranslationFunction } from 'i18next';
import { IconNames } from '../../shared/utils/iconNames';
import { ResourceKeys } from '../../../localization/resourceKeys';

export interface AzureSubscriptionsMenuBarProps {
  onRefreshClick: () => void;
  refreshDisabled: boolean;
  t: TranslationFunction;
}

export class AzureSubscriptionsMenuBar extends React.Component<AzureSubscriptionsMenuBarProps, {}> {
    public render(): JSX.Element {
        const refreshButtonProps = this.generateRefreshButtonProps();

        return (
            <CommandBar
                items={[refreshButtonProps]}
                overflowItems={[]}
                farItems={[]}
                ariaLabel={this.props.t(ResourceKeys.azureSubscriptions.commandBar.ariaLabel)}
            />
        );
    }

    private generateRefreshButtonProps(): ICommandBarItemProps {
        return {
            ariaLabel: this.props.t(ResourceKeys.azureSubscriptions.commandBar.refresh.ariaLabel),
            disabled: this.props.refreshDisabled,
            iconProps: {
                iconName: IconNames.Refresh
            },
            key: 'refresh',
            name: this.props.t(ResourceKeys.common.refresh),
            onClick: this.refreshClickHandler
        };
    }

    private refreshClickHandler = () => {
        this.props.onRefreshClick();
    }
}