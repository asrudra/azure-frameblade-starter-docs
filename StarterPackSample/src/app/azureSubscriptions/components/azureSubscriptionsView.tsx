import * as React from 'react';
import { TranslationFunction } from 'i18next';
import { IColumn, DefaultButton, DetailsList, CheckboxVisibility, MessageBar, MessageBarType } from 'office-ui-fabric-react';
import { LocalizationContextConsumer, LocationContextInterface } from '../../contexts/localizationContext';
import { AzureSubscription } from '../../services/models/azureSubscription';
import { AzureSubscriptionsMenuBar } from './azureSubscriptionsMenuBar';
import { ResourceKeys } from '../../../localization/resourceKeys';

export interface AzureSubscriptionsViewProps {
    azureSubscriptions: AzureSubscription[];
    errorFetchingAzureSubscriptions: boolean;
    fetchAzureSubscriptions: (nextLink: string) => void;
    hasFetchedAzureSubscriptions: boolean;
    isFetchingAzureSubscriptions: boolean;
    nextLink: string;
}

export class AzureSubscriptionsView extends React.Component<AzureSubscriptionsViewProps, {}> {

    constructor(props: AzureSubscriptionsViewProps) {
        super(props);
        this.state = {};
    }

    public componentDidMount() {
        if (this.props.isFetchingAzureSubscriptions || this.props.hasFetchedAzureSubscriptions) {
                return;
        }

        this.props.fetchAzureSubscriptions('');
    }

    private createListColumn = (name: string, localizationKey: string, t: TranslationFunction) => {
        return {
            fieldName: name,
            key: name,
            maxWidth: 300,
            minWidth: 100,
            name: t(localizationKey),
        };
    }

    public generateSubscriptionListColumns(t: TranslationFunction): IColumn[] {
        return [
            this.createListColumn('id', ResourceKeys.azureSubscriptions.subscriptionsTable.id, t),
            this.createListColumn('displayName', ResourceKeys.azureSubscriptions.subscriptionsTable.displayName, t),
            this.createListColumn('state', ResourceKeys.azureSubscriptions.subscriptionsTable.state, t),
            this.createListColumn('authorizationSource', ResourceKeys.azureSubscriptions.subscriptionsTable.authorizationSource, t)
        ];
    }

    public render(): JSX.Element {
        const { errorFetchingAzureSubscriptions, isFetchingAzureSubscriptions, nextLink } = this.props;

        return (
            <LocalizationContextConsumer>
                {(context: LocationContextInterface) => (
                    <div className="listViewDisplay">
                        <div className="listViewDisplay-header">
                            <AzureSubscriptionsMenuBar
                                onRefreshClick={this.refreshHandler}
                                refreshDisabled={isFetchingAzureSubscriptions}
                            />

                            {errorFetchingAzureSubscriptions &&
                                <MessageBar
                                    className="listViewDisplay-message"
                                    messageBarType={MessageBarType.warning}
                                >
                                    {context.t(ResourceKeys.azureSubscriptions.errors.failedToGetSubscriptions)}
                                </MessageBar>
                            }
                        </div>

                        <div className="listViewDisplay-body">
                            <div className="listViewDisplay-content">
                                <p>
                                    {context.t(ResourceKeys.azureSubscriptions.header.description)}
                                </p>
                            </div>

                            <div className="listViewDisplay-content">
                                <DetailsList
                                    setKey="subscriptions"
                                    checkboxVisibility={CheckboxVisibility.hidden}
                                    items={this.props.azureSubscriptions}
                                    columns={this.generateSubscriptionListColumns(context.t)}
                                />
                            </div>

                            {nextLink &&
                                <div className="listViewDisplay-content">
                                    <DefaultButton
                                        className="load-more-button"
                                        text={context.t(ResourceKeys.azureSubscriptions.subscriptionsTable.loadMore.text)}
                                        disabled={!this.props.nextLink}
                                        onClick={this.loadMoreHandler}
                                    />
                                </div>
                            }
                        </div>
                    </div>
                )}
            </LocalizationContextConsumer>

        );
    }

    private loadMoreHandler = (): void => {
        this.props.fetchAzureSubscriptions(this.props.nextLink);
    }

    private refreshHandler = (): void => {
        this.props.fetchAzureSubscriptions('');
    }
}
