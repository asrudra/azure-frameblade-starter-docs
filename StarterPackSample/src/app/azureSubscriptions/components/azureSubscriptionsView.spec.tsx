import * as React from 'react';
import { AzureSubscriptionsView } from './azureSubscriptionsView';
import { testWithLocalizationContext } from '../../shared/utils/testHelpers';
import { mount } from 'enzyme';
import { AzureSubscriptionsMenuBar } from './azureSubscriptionsMenuBar';

describe('azureSubscriptionsView', () => {

    context('snapshot', () => {
        it('matches snapshot when no error present', () => {
            const wrapper = testWithLocalizationContext(
                <AzureSubscriptionsView
                    azureSubscriptions={[]}
                    fetchAzureSubscriptions={jest.fn()}
                    hasFetchedAzureSubscriptions={false}
                    isFetchingAzureSubscriptions={false}
                    errorFetchingAzureSubscriptions={false}
                    nextLink={''}
                />);

            expect(wrapper).toMatchSnapshot();
        });

        it('matches snapshot when error present', () => {
            const wrapper = testWithLocalizationContext(
                <AzureSubscriptionsView
                    azureSubscriptions={[]}
                    fetchAzureSubscriptions={jest.fn()}
                    hasFetchedAzureSubscriptions={false}
                    isFetchingAzureSubscriptions={false}
                    errorFetchingAzureSubscriptions={true}
                    nextLink={''}
                />);

            expect(wrapper).toMatchSnapshot();
        });
    });

    context('componentDidMount', () => {
        it('calls fetch if hasFetched equals false and isFetching equals false', () => {
            const mockFetch = jest.fn();
            testWithLocalizationContext(
                <AzureSubscriptionsView
                    azureSubscriptions={[]}
                    fetchAzureSubscriptions={mockFetch}
                    hasFetchedAzureSubscriptions={false}
                    isFetchingAzureSubscriptions={false}
                    errorFetchingAzureSubscriptions={false}
                    nextLink={''}
                />);

            expect(mockFetch).toHaveBeenCalledWith('');
        });

        it('does not call fetch if hasFetched equals true', () => {
            const mockFetch = jest.fn();
            testWithLocalizationContext(
                <AzureSubscriptionsView
                    azureSubscriptions={[]}
                    fetchAzureSubscriptions={mockFetch}
                    hasFetchedAzureSubscriptions={true}
                    isFetchingAzureSubscriptions={false}
                    errorFetchingAzureSubscriptions={false}
                    nextLink={''}
                />);

            expect(mockFetch).not.toHaveBeenCalled();
        });

        it ('does not call fetch if isFetching equals true', () => {
            const mockFetch = jest.fn();
            testWithLocalizationContext(
                <AzureSubscriptionsView
                    azureSubscriptions={[]}
                    fetchAzureSubscriptions={mockFetch}
                    hasFetchedAzureSubscriptions={false}
                    isFetchingAzureSubscriptions={true}
                    errorFetchingAzureSubscriptions={false}
                    nextLink={''}
                />);

            expect(mockFetch).not.toHaveBeenCalled();
        });
    });

    context('render/load-more-button', () => {
        it('calls fetchAzureSubscriptions with next link on loadMore button click', () => {
            const mockFetch = jest.fn();
            const wrapper = testWithLocalizationContext(
                <AzureSubscriptionsView
                    azureSubscriptions={[]}
                    fetchAzureSubscriptions={mockFetch}
                    hasFetchedAzureSubscriptions={false}
                    isFetchingAzureSubscriptions={true}
                    errorFetchingAzureSubscriptions={false}
                    nextLink={'nextLink'}
                />);

            wrapper.find('.load-more-button').simulate('click');
            expect(mockFetch).toHaveBeenCalledWith('nextLink');
        });
    });

    context ('render/AzureSubscriptionsMenuBar', () => {
        it('calls fetchAzureSubscriptions with empty next link on refresh click', () => {
            const mockFetch = jest.fn();
            const wrapper = testWithLocalizationContext(
                <AzureSubscriptionsView
                    azureSubscriptions={[]}
                    fetchAzureSubscriptions={mockFetch}
                    hasFetchedAzureSubscriptions={false}
                    isFetchingAzureSubscriptions={true}
                    errorFetchingAzureSubscriptions={false}
                    nextLink={'nextLink'}
                />);

            wrapper.find(AzureSubscriptionsMenuBar).prop('onRefreshClick')();
            expect(mockFetch).toHaveBeenCalledWith('');
        });
    });
});
