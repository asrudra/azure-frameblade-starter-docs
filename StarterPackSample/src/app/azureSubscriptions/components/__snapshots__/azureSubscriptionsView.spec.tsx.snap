// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`azureSubscriptionsView snapshot matches snapshot when error present 1`] = `
<div
  className="listViewDisplay"
>
  <div
    className="listViewDisplay-header"
  >
    <AzureSubscriptionsMenuBar
      onRefreshClick={[Function]}
      refreshDisabled={false}
    />
    <StyledMessageBarBase
      className="listViewDisplay-message"
      messageBarType={5}
    />
  </div>
  <div
    className="listViewDisplay-body"
  >
    <div
      className="listViewDisplay-content"
    >
      <p />
    </div>
    <div
      className="listViewDisplay-content"
    >
      <StyledWithViewportComponent
        checkboxVisibility={2}
        columns={
          Array [
            Object {
              "fieldName": "id",
              "key": "id",
              "maxWidth": 300,
              "minWidth": 100,
              "name": undefined,
            },
            Object {
              "fieldName": "displayName",
              "key": "displayName",
              "maxWidth": 300,
              "minWidth": 100,
              "name": undefined,
            },
            Object {
              "fieldName": "state",
              "key": "state",
              "maxWidth": 300,
              "minWidth": 100,
              "name": undefined,
            },
            Object {
              "fieldName": "authorizationSource",
              "key": "authorizationSource",
              "maxWidth": 300,
              "minWidth": 100,
              "name": undefined,
            },
          ]
        }
        items={Array []}
        setKey="subscriptions"
      />
    </div>
  </div>
</div>
`;

exports[`azureSubscriptionsView snapshot matches snapshot when no error present 1`] = `
<div
  className="listViewDisplay"
>
  <div
    className="listViewDisplay-header"
  >
    <AzureSubscriptionsMenuBar
      onRefreshClick={[Function]}
      refreshDisabled={false}
    />
  </div>
  <div
    className="listViewDisplay-body"
  >
    <div
      className="listViewDisplay-content"
    >
      <p />
    </div>
    <div
      className="listViewDisplay-content"
    >
      <StyledWithViewportComponent
        checkboxVisibility={2}
        columns={
          Array [
            Object {
              "fieldName": "id",
              "key": "id",
              "maxWidth": 300,
              "minWidth": 100,
              "name": undefined,
            },
            Object {
              "fieldName": "displayName",
              "key": "displayName",
              "maxWidth": 300,
              "minWidth": 100,
              "name": undefined,
            },
            Object {
              "fieldName": "state",
              "key": "state",
              "maxWidth": 300,
              "minWidth": 100,
              "name": undefined,
            },
            Object {
              "fieldName": "authorizationSource",
              "key": "authorizationSource",
              "maxWidth": 300,
              "minWidth": 100,
              "name": undefined,
            },
          ]
        }
        items={Array []}
        setKey="subscriptions"
      />
    </div>
  </div>
</div>
`;
